File: apps/mobile/ios/Visvine/Realtime/ChatEventParser.swift
import Foundation

/// One `data:` payload of the agent chat stream → a `ChatEvent`. The server
/// writes `{ "type": … }` (lib/agents/shared/chat.ts#ChatStreamEvent); a type
/// this build does not know, or a payload that is not JSON, is nil and skipped.
enum ChatEventParser {
    private static let decoder = JSONDecoder()

    static func parse(_ data: String) -> ChatEvent? {
        guard let bytes = data.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: bytes) as? [String: Any],
              let type = obj["type"] as? String else { return nil }
        switch type {
        case "user", "done":
            guard let raw = obj["message"],
                  let payload = try? JSONSerialization.data(withJSONObject: raw),
                  let message = try? decoder.decode(ChatMessage.self, from: payload) else { return nil }
            return type == "user" ? .user(message) : .done(message)
        case "tool":
            return .tool(name: obj["tool"] as? String ?? "", detail: obj["detail"] as? String ?? "")
        case "tool_result":
            return .toolResult(name: obj["tool"] as? String ?? "", text: obj["text"] as? String ?? "")
        case "assistant":
            return .assistant(text: obj["text"] as? String ?? "")
        case "error":
            return .error(reason: obj["reason"] as? String ?? "error", message: obj["message"] as? String ?? "Something went wrong.")
        default:
            return nil
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/visvine/mobile/data/realtime/ChatEventParser.kt
package com.visvine.mobile.data.realtime

import com.visvine.mobile.data.model.ChatEvent
import com.visvine.mobile.data.model.ChatMessage
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

/**
 * One `data:` line of `POST …/agents/<name>/chat/stream` → a [ChatEvent], or
 * null for a type this client does not know or a payload that does not parse.
 * Pure — the stream and the test both call it. The server names the event in
 * the JSON's `type` field, never in an SSE `event:` line
 * (lib/agents/shared/chat.ts#ChatStreamEvent).
 */
fun parseChatEvent(json: Json, data: String): ChatEvent? {
    val obj = sseDataObject(json, data) ?: return null
    val type = obj.string("type") ?: return null
    return runCatching {
        when (type) {
            "user" -> ChatEvent.User(obj.message(json) ?: return null)
            "tool" -> ChatEvent.Tool(tool = obj.string("tool") ?: "", detail = obj.string("detail") ?: "")
            "tool_result" -> ChatEvent.ToolResult(tool = obj.string("tool") ?: "", text = obj.string("text") ?: "")
            "assistant" -> ChatEvent.Assistant(text = obj.string("text") ?: "")
            "done" -> ChatEvent.Done(obj.message(json) ?: return null)
            "error" -> ChatEvent.Error(reason = obj.string("reason") ?: "error", message = obj.string("message") ?: "")
            else -> null
        }
    }.getOrNull()
}

private fun JsonObject.string(key: String): String? =
    runCatching { this[key]?.jsonPrimitive?.content }.getOrNull()

private fun JsonObject.message(json: Json): ChatMessage? =
    runCatching { json.decodeFromJsonElement(ChatMessage.serializer(), this["message"]!!.jsonObject) }.getOrNull()
